pkg/webhook/resources/virtualmachineimage/mutator.go

```diff
@@ -0,0 +1,116 @@
+package virtualmachineimage
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	longhorntypes "github.com/longhorn/longhorn-manager/types"
+	ctlstoragev1 "github.com/rancher/wrangler/pkg/generated/controllers/storage/v1"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	"github.com/harvester/harvester/pkg/util"
+	"github.com/harvester/harvester/pkg/webhook/types"
+)
+
+func NewMutator(storageClassCache ctlstoragev1.StorageClassCache) types.Mutator {
+	return &virtualMachineImageMutator{
+		storageClassCache: storageClassCache,
+	}
+}
+
+type virtualMachineImageMutator struct {
+	types.DefaultMutator
+	storageClassCache ctlstoragev1.StorageClassCache
+}
+
+func (m *virtualMachineImageMutator) Resource() types.Resource {
+	return types.Resource{
+		Names:      []string{harvesterv1.VirtualMachineImageResourceName},
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   harvesterv1.SchemeGroupVersion.Group,
+		APIVersion: harvesterv1.SchemeGroupVersion.Version,
+		ObjectType: &harvesterv1.VirtualMachineImage{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+		},
+	}
+}
+
+func (m *virtualMachineImageMutator) Create(request *types.Request, newObj runtime.Object) (types.PatchOps, error) {
+	newImage := newObj.(*harvesterv1.VirtualMachineImage)
+
+	return m.patchImageStorageClassParams(newImage)
+}
+
+func (m *virtualMachineImageMutator) patchImageStorageClassParams(newImage *harvesterv1.VirtualMachineImage) ([]string, error) {
+	var patchOps types.PatchOps
+
+	storageClassName := newImage.Annotations[util.AnnotationStorageClassName]
+	storageClass, err := m.getStorageClass(storageClassName)
```

Comment:
storage class of backing images is a special design in Harvester and I believe its `classParams` like replicas, stale should be consistent across all images. Storage tiers should not apply to the backing images as any VM images should be utilized to create a VM in any nodes. So I would prefer to:
1. create a particular default storage class for the backing images, and it can be modified via the `kubectl` only when needed (but not exposed in the UI by default).
2. define those default values in the Harvester code.

Reply:
the case is using different diskSelectors in different images

Reply:
sure, then as long as we use an appropriate default SC it should be ok, thanks.